File: py/yeriasdk/core/yeria_link.py
"""Canonical Yeria deeplink generator."""

import re
from typing import Literal
from urllib.parse import parse_qs, quote, unquote, urlparse

from ..errors.exceptions import InvalidParameterError

YeriaLinkFormat = Literal["https", "yeria"]


class YeriaLink:
    """Generate links that open a destination in Yeria."""

    _SERVICE_ID_PATTERN = r"^[A-Za-z0-9._~-]{1,128}$"

    @classmethod
    def service(cls, service_id: str, format: YeriaLinkFormat = "https") -> str:
        return cls._base(service_id, "s", format)

    @classmethod
    def component(
        cls,
        service_id: str,
        provider_path: str,
        format: YeriaLinkFormat = "https",
    ) -> str:
        path = cls._assert_provider_path(provider_path)
        return f"{cls._base(service_id, 'v', format)}?p={quote(path, safe='')}"

    @classmethod
    def chat(cls, service_id: str, format: YeriaLinkFormat = "https") -> str:
        return cls._base(service_id, "c", format)

    @classmethod
    def pin(cls, service_id: str, format: YeriaLinkFormat = "https") -> str:
        return cls._base(service_id, "p", format)

    @classmethod
    def subscribe(cls, service_id: str, format: YeriaLinkFormat = "https") -> str:
        return cls._base(service_id, "n", format)

    @classmethod
    def is_valid(cls, link: str) -> bool:
        try:
            trimmed = link.strip()
            if (
                not trimmed
                or not trimmed.lower().startswith(("https://", "yeria://"))
                or "\\" in trimmed
                or any(ord(char) <= 31 or ord(char) == 127 for char in trimmed)
            ):
                return False

            parsed = urlparse(trimmed)
            if parsed.fragment or parsed.username or parsed.password:
                return False
            if cls._has_noncanonical_route_path(parsed.path):
                return False

            if parsed.scheme == "yeria":
                if parsed.hostname != "dl" or parsed.port is not None:
                    return False
                segments = [part for part in parsed.path.split("/") if part]
            elif parsed.scheme == "https":
                if parsed.hostname != "yeria.app" or parsed.port not in (None, 443):
                    return False
                segments = [part for part in parsed.path.split("/") if part]
                if not segments or segments.pop(0) != "dl":
                    return False
            else:
                return False

            if len(segments) != 2:
                return False
            operation, service_id = segments
            if not cls._is_service_id(service_id):
                return False

            if operation == "s":
                return not parsed.query

            if operation == "v":
                query = parse_qs(parsed.query, keep_blank_values=True)
                paths = query.get("p")
                return (
                    len(query) == 1
                    and paths is not None
                    and len(paths) == 1
                    and cls._is_provider_path(paths[0])
                )

            return operation in ("c", "p", "n") and not parsed.query
        except ValueError:
            return False

    @staticmethod
    def _has_noncanonical_route_path(path: str) -> bool:
        if not path.startswith("/"):
            return True

        raw_segments = path[1:].split("/")
        if not raw_segments or any(not segment for segment in raw_segments):
            return True

        return any(unquote(segment) in (".", "..") for segment in raw_segments)

    @classmethod
    def _base(
        cls,
        service_id: str,
        operation: Literal["s", "p", "n", "c", "v"],
        format: YeriaLinkFormat,
    ) -> str:
        service_id = service_id.strip()
        if not cls._is_service_id(service_id):
            raise InvalidParameterError(
                "service_id",
                service_id,
                "Service ID must contain 1 to 128 URL-safe characters",
            )

        if format == "https":
            return f"https://yeria.app/dl/{operation}/{service_id}"
        if format == "yeria":
            return f"yeria://dl/{operation}/{service_id}"
        raise InvalidParameterError(
            "format", format, "Format must be 'https' or 'yeria'"
        )

    @classmethod
    def _assert_provider_path(cls, provider_path: str) -> str:
        path = provider_path.strip()
        if not cls._is_provider_path(path):
            raise InvalidParameterError(
                "provider_path",
                provider_path,
                "Provider path must be relative, non-empty, and contain no traversal or fragment",
            )
        return path

    @classmethod
    def _is_service_id(cls, service_id: str) -> bool:
        return re.fullmatch(cls._SERVICE_ID_PATTERN, service_id) is not None

    @staticmethod
    def _is_provider_path(path: str) -> bool:
        if not path or path.startswith("//") or "\\" in path or "#" in path:
            return False

        try:
            parsed = urlparse(path)
            if parsed.scheme or parsed.netloc or not parsed.path:
                return False
            return not any(
                unquote(segment) in (".", "..")
                for segment in parsed.path.split("/")
            )
        except ValueError:
            return False

File: js/src/core/yeria-link.ts
import { InvalidParameterError } from '../errors';

export type YeriaLinkFormat = 'https' | 'yeria';

export interface YeriaLinkOptions {
    format?: YeriaLinkFormat;
}

/**
 * Generates canonical links that open a destination in Yeria.
 *
 * HTTPS links are the default because they work on web pages as well as on
 * devices with Yeria installed. The custom `yeria:` scheme is available for
 * environments that already know the application is installed.
 */
export class YeriaLink {
    private static readonly serviceIdPattern = /^[A-Za-z0-9._~-]{1,128}$/;

    static service(serviceId: string, options: YeriaLinkOptions = {}): string {
        return this.base(serviceId, 's', options);
    }

    static component(
        serviceId: string,
        providerPath: string,
        options: YeriaLinkOptions = {}
    ): string {
        const path = this.assertProviderPath(providerPath);
        return `${this.base(serviceId, 'v', options)}?p=${encodeURIComponent(path)}`;
    }

    static chat(serviceId: string, options: YeriaLinkOptions = {}): string {
        return this.base(serviceId, 'c', options);
    }

    static pin(serviceId: string, options: YeriaLinkOptions = {}): string {
        return this.base(serviceId, 'p', options);
    }

    static subscribe(serviceId: string, options: YeriaLinkOptions = {}): string {
        return this.base(serviceId, 'n', options);
    }

    static isValid(link: string): boolean {
        const trimmed = link.trim();
        if (!trimmed ||
            !/^(?:https|yeria):\/\//i.test(trimmed) ||
            /[\u0000-\u001F\u007F\\]/.test(trimmed) ||
            this.hasNonCanonicalRoutePath(trimmed)) {
            return false;
        }

        let url: URL;
        try {
            url = new URL(trimmed);
        } catch {
            return false;
        }

        if (url.hash || url.username || url.password) return false;

        let segments: string[];
        if (url.protocol === 'yeria:') {
            if (url.hostname.toLowerCase() !== 'dl' || url.port) return false;
            segments = url.pathname.split('/').filter(Boolean);
        } else if (url.protocol === 'https:') {
            if (url.hostname.toLowerCase() !== 'yeria.app' || url.port) return false;
            segments = url.pathname.split('/').filter(Boolean);
            if (segments.shift() !== 'dl') return false;
        } else {
            return false;
        }

        if (segments.length !== 2) return false;
        const [operation, serviceId] = segments;
        if (!serviceId || !this.serviceIdPattern.test(serviceId)) return false;

        if (operation === 's') {
            return url.search === '';
        }

        if (operation === 'v') {
            const keys = [...url.searchParams.keys()];
            const paths = url.searchParams.getAll('p');
            const providerPath = paths[0];
            return keys.length === 1 &&
                keys[0] === 'p' &&
                paths.length === 1 &&
                providerPath !== undefined &&
                this.isProviderPath(providerPath);
        }

        return operation !== undefined &&
            ['c', 'p', 'n'].includes(operation) &&
            url.search === '';
    }

    private static hasNonCanonicalRoutePath(link: string): boolean {
        const authorityStart = link.indexOf('://') + 3;
        const pathStart = link.indexOf('/', authorityStart);
        if (pathStart < 0) return true;

        const queryStart = link.indexOf('?', pathStart);
        const fragmentStart = link.indexOf('#', pathStart);
        const pathEnd = [queryStart, fragmentStart]
            .filter(index => index >= 0)
            .reduce((earliest, index) => Math.min(earliest, index), link.length);
        const rawSegments = link.slice(pathStart + 1, pathEnd).split('/');

        if (rawSegments.length === 0 || rawSegments.some(segment => segment === '')) {
            return true;
        }

        try {
            return rawSegments.some(segment => {
                const decoded = decodeURIComponent(segment);
                return decoded === '.' || decoded === '..';
            });
        } catch {
            return true;
        }
    }

    private static base(
        serviceId: string,
        operation: 's' | 'p' | 'n' | 'c' | 'v',
        options: YeriaLinkOptions
    ): string {
        const id = serviceId.trim();
        if (!this.serviceIdPattern.test(id)) {
            throw new InvalidParameterError(
                'serviceId',
                serviceId,
                'Service ID must contain 1 to 128 URL-safe characters'
            );
        }

        const format = options.format ?? 'https';
        if (format === 'https') return `https://yeria.app/dl/${operation}/${id}`;
        if (format === 'yeria') return `yeria://dl/${operation}/${id}`;
        throw new InvalidParameterError('format', format, "Format must be 'https' or 'yeria'");
    }

    private static assertProviderPath(providerPath: string): string {
        const path = providerPath.trim();
        if (!this.isProviderPath(path)) {
            throw new InvalidParameterError(
                'providerPath',
                providerPath,
                'Provider path must be relative, non-empty, and contain no traversal or fragment'
            );
        }
        return path;
    }

    private static isProviderPath(path: string): boolean {
        if (!path || path.startsWith('//') || path.includes('\\') || path.includes('#')) {
            return false;
        }

        try {
            const rawPath = path.split('?')[0] ?? '';
            if (!rawPath) return false;
            if (rawPath.split('/').some(segment => {
                const decoded = decodeURIComponent(segment);
                return decoded === '.' || decoded === '..';
            })) {
                return false;
            }

            const parsed = new URL(path, 'https://service.invalid');
            if (parsed.origin !== 'https://service.invalid' || !parsed.pathname) return false;
            return true;
        } catch {
            return false;
        }
    }
}
